rmf_traffic/src/rmf_traffic/schedule/Timeline.hpp

```diff
@@ -0,0 +1,517 @@
+    }
+  }
+
+  void cull(const Time time)
+  {
+    for (auto& pair : _timelines)
+    {
+      auto& timeline = pair.second;
+
+      // NOTE(MXG): It is not an error that we are using get_timeline_begin() to
+      // find the ending iterator. We want to stop just before the first bucket
+      // that contains the cull time, because we only want to erase times that
+      // come before it.
+      const auto end_it = get_timeline_begin(timeline, &time);
+      if (end_it != timeline.begin())
+        timeline.erase(timeline.begin(), end_it);
+    }
+  }
+
+private:
+
+  template<typename Inspector, typename ParticipantFilter>
+  void inspect_spacetime(
+      const Query::Spacetime& spacetime,
+      const ParticipantFilter& participant_filter,
+      Inspector& inspector) const
+  {
+    const Query::Spacetime::Mode mode = spacetime.get_mode();
+
+    if (Query::Spacetime::Mode::All == mode)
+    {
+      inspect_all_spacetime(participant_filter, inspector);
+    }
+    else if (Query::Spacetime::Mode::Regions == mode)
+    {
+      inspect_spacetime_regions(
+            spacetime.regions(), participant_filter, inspector);
+    }
+    else if (Query::Spacetime::Mode::Timespan == mode)
+    {
+      inspect_spacetime_timespan(
+            spacetime.timespan(), participant_filter, inspector);
+    }
+  }
+
+  template<typename Inspector, typename ParticipantFilter>
+  void inspect_all_spacetime(
+      const ParticipantFilter& participant_filter,
+      Inspector& inspector) const
+  {
+    Checked checked;
+
+    const auto relevant = [](const Route&) -> bool { return true; };
+    for (const auto& entry : _all_bucket)
+    {
+      if (participant_filter.ignore(entry->participant))
+        continue;
+
+      if (!checked[entry->participant].insert(entry->route_id).second)
+        continue;
+
+      inspector.inspect(entry, relevant);
+    }
+  }
+
+  template<typename Inspector, typename ParticipantFilter>
+  void inspect_spacetime_regions(
+      const Query::Spacetime::Regions& regions,
+      const ParticipantFilter& participant_filter,
+      Inspector& inspector) const
+  {
+    Checked checked;
+
+    rmf_traffic::internal::Spacetime spacetime_data;
+    const auto relevant = [&spacetime_data](const Route& r) -> bool {
+      return rmf_traffic::internal::detect_conflicts(
+            r.trajectory(), spacetime_data, nullptr);
+    };
+
+    for (const Region& region : regions)
+    {
+      const std::string& map = region.get_map();
+      const auto map_it = _timelines.find(map);
+      if (map_it == _timelines.end())
+        continue;
+
+      const Entries& timeline = map_it->second;
+      const Time* const lower_time_bound = region.get_lower_time_bound();
+      const Time* const upper_time_bound = region.get_upper_time_bound();
+
+      spacetime_data.lower_time_bound = lower_time_bound;
+      spacetime_data.upper_time_bound = upper_time_bound;
+
+      const auto timeline_begin =
```

Comment:
I'm guessing this was written before a refactor.
```suggestion
      const auto timeline_begin = get_timeline_begin(timeline. lower_time_bound);
```